HPCC-15037 Add mechanism to detach ESP from DALI

rpastrana opened the pull request:
Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

rpastrana commented:
some comment history: https://github.com/hpcc-systems/HPCC-Platform/pull/10469

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: f5bcd45b9110dbd6befa2573ad23ef13bf17c382
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |81 |81 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |740 |740 |0 |
|test (thor) |649 |649 |0 |
|test (roxie) |767 |767 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   26 sec (00:00:26) |  178 sec (00:02:58) |   56 sec (00:00:56) |    5 sec (00:00:05) |   33 sec (00:00:33) | 1099 sec (00:18:19) |   33 sec (00:00:33) | 1430 sec (00:23:50) |

rpastrana commented:
@afishbeck please review

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 95c5721791c75854dbdbfb9aaf71ef6f9a77692e
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |81 |81 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |741 |741 |0 |
|test (thor) |650 |650 |0 |
|test (roxie) |768 |768 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   13 sec (00:00:13) |  180 sec (00:03:00) |   56 sec (00:00:56) |    6 sec (00:00:06) |   34 sec (00:00:34) | 1098 sec (00:18:18) |   32 sec (00:00:32) | 1419 sec (00:23:39) |

rpastrana commented:
@afishbeck  bump

afishbeck commented on esp/bindings/SOAP/Platform/soapbind.hpp:
```diff
@@ -220,6 +220,31 @@ class esp_http_decl CHttpSoapBinding : public CSoapBinding, public EspHttpBindin
 
+    virtual bool subscribeBindingToDali()
+    {
+        //can a binding have more than 1 service? if so, why is there a getService()?
```
Yes, an ESPbinding only has 1 ESPservice.  getSerice() should return that one service.
Not to be confused with something like DynamicESDL that has virtual bindings and services.
Note: An ESP process can have multiple ports, each port can run multiple bindings.. each binding has one service but that can be many within a given process.

You should remove these comments once you are satisfied with the answer.

afishbeck commented on esp/platform/espcfg.cpp:
```diff
@@ -272,10 +276,45 @@ CEspConfig::CEspConfig(IProperties* inputs, IPropertyTree* envpt, IPropertyTree*
+        {
+            StringBuffer procDirectory;
+            m_cfg->getProp("@directory", procDirectory);
+            m_esdlStateFilesLocation.setf("%s%c%s-AttachState.xml",procDirectory.str(), PATHSEPCHAR, m_process.str());
```
should you handle the case where directory is not defined?

afishbeck commented on esp/platform/espcfg.cpp:
```diff
@@ -272,10 +276,45 @@ CEspConfig::CEspConfig(IProperties* inputs, IPropertyTree* envpt, IPropertyTree*
+                Owned<IPTree> espProcAttachState = createPTreeFromXMLFile(m_esdlStateFilesLocation);
+                if (espProcAttachState)
+                {
+                    const char * restoredBindingTS = espProcAttachState->queryProp("@StateSaveTimems");
```
Is "restoredBindingTS" used?

afishbeck commented on esp/platform/espcfg.cpp:
```diff
@@ -272,10 +276,45 @@ CEspConfig::CEspConfig(IProperties* inputs, IPropertyTree* envpt, IPropertyTree*
+                {
+                    const char * restoredBindingTS = espProcAttachState->queryProp("@StateSaveTimems");
+                    setDetachedFromDali(!(espProcAttachState->getPropBool("@attached", true)));
+                    bool subscribed = espProcAttachState->getPropBool("@subscribed", true);
```
is "subscribed" used?

afishbeck commented on esp/platform/espcfg.cpp:
```diff
@@ -272,10 +276,45 @@ CEspConfig::CEspConfig(IProperties* inputs, IPropertyTree* envpt, IPropertyTree*
+        if(!isDetachedFromDali())
+        {
+            StringBuffer daliservers;
+            if (m_cfg->getProp("@daliServers", daliservers))
```
Haven't looked closely, but can you optimize the case where there is no dali server configured?  I.e. this ESP is configured to be completely detached from DALI.

afishbeck commented on esp/platform/espcfg.ipp:
```diff
@@ -202,14 +234,28 @@ public:
+        }
+        catch(IException* ie)
+        {
+            if (isDetachedFromDali())
```
Should you move this check to the top and not even try?  Otherwise maybe change the wording a little bit.  The exception could be for something else, in which case being detached from DALI is ALSO an error.  (Picky, but as simple as changing wording from "Could not" to "Can not".)

afishbeck replied:
@rpastrana ?

rpastrana replied:
The problem is that some ESP services can load while Detached from DALI, this message conveys the fact that in this particular system, we were not able to load ESP service(s) while in detached state. We don't know if the failure is due to the detached state either.
Perhaps rewording as such could be more accurate and helpful:
"Could not load ESP service(s) while DETACHED from DALI - Consider re-attaching ESP process"

afishbeck commented on esp/scm/ws_espcontrol.ecm:
```diff
@@ -93,13 +93,59 @@ ESPresponse [exceptions_inline, nil_remove, http_encode(0)] SetSessionTimeoutRes
+    string Message;
+};
+
+ESPrequest [nil_remove] ReSubscribeToDaliRequest
```
I'm not sure I like the "Re" part of the name.  I.e. you "Subscribe", "Unsubscribe", and then "Subscribe" again.

afishbeck commented on esp/services/esdl_svc_engine/esdl_binding.hpp:
```diff
@@ -303,9 +330,11 @@ class EsdlBindingImpl : public CHttpSoapBinding
+    CriticalSection                         detachCritSec;
     StringBuffer                            m_esdlStateFilesLocation;
-    MapStringTo<SecAccessFlags>             m_accessmap;
+    //MapStringTo<SecAccessFlags>             m_accessmap;
```
Don't add commented out code?

afishbeck commented on esp/services/esdl_svc_engine/esdl_binding.hpp:
```diff
@@ -381,6 +410,101 @@ class EsdlBindingImpl : public CHttpSoapBinding
+    }
+
+/*
+    virtual bool attach()
```
Need this commented out code?  Was it basically moved?

afishbeck commented on esp/services/espcontrol/ws_espcontrolservice.cpp:
```diff
@@ -75,6 +75,104 @@ void CWSESPControlEx::init(IPropertyTree *cfg, const char *process, const char *
+{
+    int status = 0;
+    bool force = req.getForce_isNull() ? false : req.getForce();
+    VStringBuffer message("Request to DETACH ESP Process '%s' from Dali ", this->espProcess.get());
```
Why the need for this->?

afishbeck commented on esp/services/espcontrol/ws_espcontrolservice.cpp:
```diff
@@ -75,6 +75,104 @@ void CWSESPControlEx::init(IPropertyTree *cfg, const char *process, const char *
+bool CWSESPControlEx::onAttachToDali(IEspContext& context, IEspAttachToDaliRequest& req, IEspAttachToDaliResponse& resp)
+{
+    int status = 0;
+    VStringBuffer message("Request to ATTACH ESP Process '%s' to Dali ", this->espProcess.get());
```
Why the need for this->?

afishbeck commented on esp/services/espcontrol/ws_espcontrolservice.cpp:
```diff
@@ -75,6 +75,104 @@ void CWSESPControlEx::init(IPropertyTree *cfg, const char *process, const char *
+}
+
+bool CWSESPControlEx::onAttachToDali(IEspContext& context, IEspAttachToDaliRequest& req, IEspAttachToDaliResponse& resp)
+{
```
Looks like you could write a common helper function to common up most of the code in onAttachToDali and onDetachFromDali?

afishbeck replied:
@rpastrana ?

rpastrana replied:
Take a look at handleDaliAttachmentRequest, if that's not what you intended, please let me know what you are envisioning.

afishbeck commented on esp/services/espcontrol/ws_espcontrolservice.cpp:
```diff
@@ -75,6 +75,104 @@ void CWSESPControlEx::init(IPropertyTree *cfg, const char *process, const char *
+}
+
+bool CWSESPControlEx::onUnsubscribeFromDali(IEspContext& context, IEspUnsubscribeFromDaliRequest& req, IEspUnsubscribeFromDaliResponse& resp)
+{
```
Same comments as above... why this-> and could common up code.

rpastrana replied:
this-> usually comes about from utilizing intellisense to find member vars, it makes it to the pull req because it helps identifying member vars without any other context. some might see it as noise, but it can actually be helpful sans IDE colorization/etc.

afishbeck commented on esp/services/ws_ecl/ws_ecl_service.cpp:
```diff
@@ -330,13 +330,13 @@ void CWsEclBinding::getNavigationData(IEspContext &context, IPropertyTree & data
     ensureNavDynFolder(data, "Targets", "Targets", "root=true", NULL);
 }
-
+/*
```
Delete rather than comment out?

afishbeck commented on esp/services/ws_loggingservice/loggingserviceplugin.cpp:
```diff
@@ -42,7 +42,7 @@ extern "C"
     {
         if (streq(type, "loggingservice_binding"))
-            return new CWsLoggingServiceSoapBinding(cfg, name, process, hsl_none);
+            return new CWsLoggingServiceSoapBindingEx(cfg, name, process, hsl_none);
```
Not sure what this bought you, but OK.

rpastrana replied:
In one of my many iterations of this change I believe I was implementing/overriding some of the new detach/attach subscribe/unsubscribe methods, not needed now.

afishbeck commented on esp/services/ws_topology/ws_topologyService.hpp:
```diff
@@ -68,6 +68,20 @@ class CWsTopologySoapBindingEx : public CWsTopologySoapBinding
         ensureNavLink(*folder, "System Servers", "/WsTopology/TpServiceQuery?Type=ALLSERVICES", "View details about System Support Servers clusters and optionally run preflight activities", NULL, NULL, 3);
     }
+
+   /* virtual bool canDetachFromDali() override
```
New commented out code?

afishbeck commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -2990,8 +2993,10 @@ int WUSchedule::run()
-    if (m_container)
-        m_container->exitESP();
+    //Rodrigo: do we need to do this anyway?? why not attempt to iterate until dali is reachable again??
+    //but anyway do we want to notify this thread that we're attached/detached?
```
Needs investigating.

rpastrana replied:
@wangkx an accessible DALI will no longer be a requirement for ESP Processes, do you see any issue stemming from removing line 2993-2994 ?

wangkx replied:
@rpastrana if those 2 lines are removed, an alternative solution has to be added to restart the WUSchedule loop whenever dali is accessible again.

afishbeck commented on esp/services/ws_workunits/ws_workunitsService.hpp:
```diff
@@ -347,7 +365,21 @@ class CWsWorkunitsSoapBindingEx : public CWsWorkunitsSoapBinding
         CWsWorkunitsSoapBinding::addService(name, host, port, service);
     }
-
+/*
```
?

afishbeck reviewed (commented):
@rpastrana some comments.

richardkchapman commented:
Needs rebasing for a clean merge

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 0b5c42b374a77b677f1558d0dc39a7e313d38162
Number of Error(s): 21
 /usr/bin/install -c -m 644 include/antlr3.h include/antlr3baserecognizer.h include/antlr3basetree.h include/antlr3basetreeadaptor.h include/antlr3bitset.h include/antlr3collections.h include/antlr3commontoken.h include/antlr3commontree.h include/antlr3commontreeadaptor.h include/antlr3commontreenodestream.h include/antlr3convertutf.h include/antlr3cyclicdfa.h include/antlr3debugeventlistener.h include/antlr3defs.h include/antlr3encodings.h include/antlr3errors.h include/antlr3exception.h include/antlr3filestream.h include/antlr3input.h include/antlr3interfaces.h include/antlr3intstream.h include/antlr3lexer.h include/antlr3memory.h include/antlr3parser.h include/antlr3parsetree.h include/antlr3recognizersharedstate.h include/antlr3rewritestreams.h include/antlr3string.h include/antlr3tokenstream.h include/antlr3treeparser.h antlr3config.h /mnt/disk1/home/vamosax/smoketest/PR-10832/build/esp/services/ws_sql/antlr3c/include
[  0%] Building CXX object aws-cpp-sdk-core/CMakeFiles/aws-cpp-sdk-core.dir/source/client/CoreErrors.cpp.o
[100%] Building CXX object aws-cpp-sdk-sqs/CMakeFiles/aws-cpp-sdk-sqs.dir/source/SQSErrors.cpp.o
/mnt/disk1/home/vamosax/smoketest/PR-10832/HPCC-Platform/esp/services/esdl_svc_engine/esdl_binding.hpp:181:1: error: expected unqualified-id before \xe2\x80\x98<<\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10832/HPCC-Platform/esp/services/esdl_svc_engine/esdl_svc_engine.cpp:64:1: error: expected \xe2\x80\x98}\xe2\x80\x99 at end of input
/mnt/disk1/home/vamosax/smoketest/PR-10832/HPCC-Platform/esp/services/esdl_svc_engine/esdl_svc_engine.cpp:64:1: error: expected unqualified-id at end of input
make[2]: *** [esp/services/esdl_svc_engine/CMakeFiles/esdl_svc_engine.dir/esdl_svc_engine.cpp.o] Error 1
/mnt/disk1/home/vamosax/smoketest/PR-10832/HPCC-Platform/esp/services/esdl_svc_engine/esdl_binding.hpp:181:1: error: expected unqualified-id before \xe2\x80\x98<<\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10832/HPCC-Platform/esp/services/esdl_svc_engine/esdl_store.cpp:859:1: error: expected \xe2\x80\x98}\xe2\x80\x99 at end of input
/mnt/disk1/home/vamosax/smoketest/PR-10832/HPCC-Platform/esp/services/esdl_svc_engine/esdl_store.cpp:859:1: error: expected unqualified-id at end of input
/mnt/disk1/home/vamosax/smoketest/PR-10832/HPCC-Platform/esp/services/esdl_svc_engine/esdl_binding.hpp:181:1: error: expected unqualified-id before \xe2\x80\x98<<\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10832/HPCC-Platform/esp/services/esdl_svc_engine/esdl_svc_engine_plugin.cpp:49:2: error: expected \xe2\x80\x98}\xe2\x80\x99 at end of input
/mnt/disk1/home/vamosax/smoketest/PR-10832/HPCC-Platform/esp/services/esdl_svc_engine/esdl_svc_engine_plugin.cpp:49:2: error: expected unqualified-id at end of input
/mnt/disk1/home/vamosax/smoketest/PR-10832/HPCC-Platform/esp/services/esdl_svc_engine/esdl_binding.hpp:181:1: error: expected unqualified-id before \xe2\x80\x98<<\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10832/HPCC-Platform/esp/services/esdl_svc_engine/esdl_binding.cpp:2974:1: error: expected \xe2\x80\x98}\xe2\x80\x99 at end of input
/mnt/disk1/home/vamosax/smoketest/PR-10832/HPCC-Platform/esp/services/esdl_svc_engine/esdl_binding.cpp:2974:1: error: expected unqualified-id at end of input
make[2]: *** [esp/services/esdl_svc_engine/CMakeFiles/esdl_svc_engine.dir/esdl_store.cpp.o] Error 1
make[2]: *** [esp/services/esdl_svc_engine/CMakeFiles/esdl_svc_engine.dir/esdl_svc_engine_plugin.cpp.o] Error 1
make[2]: *** [esp/services
 ... ( comment is too long, 30 lines, 4171 bytes, truncated) ... 

make[1]: *** [esp/services/esdl_svc_engine/CMakeFiles/esdl_svc_engine.dir/all] Error 2
make: *** [all] Error 2
Get ECLWatch build result

[Build error:](https://gist.github.com/69275977d4a5b83df6a7f720e70d074c)
Build log [Build log](https://gist.github.com/HPCCSmoketest/69275977d4a5b83df6a7f720e70d074c#file-result-18-03-14-11-00-37-log)

```

rpastrana commented on esp/services/esdl_svc_engine/esdl_binding.hpp:
```diff
@@ -458,43 +455,50 @@ class EsdlBindingImpl : public CHttpSoapBinding, implements IEsdlListener
+            reloadBindingFromCentralStore(m_bindingName.get(), m_processName.get());
+            setIsAttached(true);
+
+            /* rodrigo
```
@mayx while rebasing your ESDLStore changes, I realized we're only subscribing to changes in /ESDL/Bindings not in /ESDL/Definitions. Was that intentional? We need to be notified of changes to the current definition.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 3f5ba37365c3bf62227bf5cea88110328ebe932e
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |83 |83 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |765 |765 |0 |
|test (thor) |679 |679 |0 |
|test (roxie) |792 |792 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   12 sec (00:00:12) |  186 sec (00:03:06) |   56 sec (00:00:56) |    5 sec (00:00:05) |   38 sec (00:00:38) | 1149 sec (00:19:09) |   32 sec (00:00:32) | 1478 sec (00:24:38) |

rpastrana commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -49,12 +49,20 @@
 int Schedule::run()
 {
-    try
+    PROGLOG("DfuWorkunit WUSchedule Thread started.");
```
@wangkx  please review wsfs schedule thread structure

rpastrana commented:
@afishbeck please review latest commit

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 8ed7e11294b94117be97f7d46d966f2f1c8c683e
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |83 |83 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |759 |759 |0 |
|test (thor) |686 |686 |0 |
|test (roxie) |799 |799 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   15 sec (00:00:15) |  187 sec (00:03:07) |   55 sec (00:00:55) |    6 sec (00:00:06) |   21 sec (00:00:21) | 1143 sec (00:19:03) |   17 sec (00:00:17) | 1444 sec (00:24:04) |

wangkx commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -2973,30 +2980,31 @@ int WUSchedule::run()
                         {
                             StringBuffer msg;
-                            ERRLOG("Exception %d:%s in WsWorkunits Schedule::run", e->errorCode(), e->errorMessage(msg).str());
+                            ERRLOG("Exception %d:%s in WsWorkunits Schedule::run while updating WU", e->errorCode(), e->errorMessage(msg).str());
```
Could throw an exception in runWorkUnit() (not only while updating).

wangkx commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -2973,30 +2980,31 @@ int WUSchedule::run()
-    //Rodrigo: do we need to do this anyway?? why not attempt to iterate until dali is reachable again??
-    //but anyway do we want to notify this thread that we're attached/detached?
-    //if (m_container)
-    //    m_container->exitESP();
```
Worth mentioning this change in RedBook?

wangkx reviewed (commented):
@rpastrana please see my comments.

rpastrana commented:
@wangkx  please review latest commit

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: a00f99ed718f4564dab0a0e49e0151b810abd2b4
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |83 |83 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |759 |759 |0 |
|test (thor) |686 |686 |0 |
|test (roxie) |799 |799 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    6 sec (00:00:06) |  180 sec (00:03:00) |   55 sec (00:00:55) |    6 sec (00:00:06) |   21 sec (00:00:21) | 1096 sec (00:18:16) |   17 sec (00:00:17) | 1381 sec (00:23:01) |

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 3f14daf7a94c1fd48ebb26e9605a72397b2bc9df
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |83 |83 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |759 |759 |0 |
|test (thor) |686 |686 |0 |
|test (roxie) |799 |799 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    5 sec (00:00:05) |  183 sec (00:03:03) |   55 sec (00:00:55) |    6 sec (00:00:06) |   22 sec (00:00:22) | 1120 sec (00:18:40) |   16 sec (00:00:16) | 1407 sec (00:23:27) |

rpastrana commented:
@afishbeck squashed changes from Kevin's code review. Please review.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: f83ebe37548039066790bfba92fc158b71ff7412
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |83 |83 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |759 |759 |0 |
|test (thor) |686 |686 |0 |
|test (roxie) |799 |799 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   32 sec (00:00:32) |  182 sec (00:03:02) |   56 sec (00:00:56) |    5 sec (00:00:05) |   21 sec (00:00:21) | 1127 sec (00:18:47) |   17 sec (00:00:17) | 1440 sec (00:24:00) |

richardkchapman commented:
Why is detaching ESP from Dali a good idea? Just curious...

rpastrana commented:
As I understand, the requirement for this feature stems from the fact that the ESP process is very sensitive to Dali's health and can shutdown if the target Dali is unresponsive.
It doesn't make much sense for eclwatch to be detached, but transactional ESP services would gain stability.

richardkchapman commented:
Needs rebasing for a clean merge

rpastrana committed:
HPCC-15037 Add mechanism to detach ESP from DALI

- Adds ESPControl methods to Attach/detach subscribe/unsubscribe from Dali
- Adds ESPServer and ESPConfig to support above methods
- Changes ESP behavior to prevent shutdown due to missing Dali connection
- Restructures esdl-binding Dali logic
- Adds unsubscribe logic to wsdfu
- Adds esdlconfig detached logic
- Adds ws_packageproecess subscribe/unsubscribe logic
- Remove ws_wuhelper shutdown of ESP if Dali not reachable

- Restructures wswu and wsfs schedule threads to withstand dali issues
- Saves ESP Proc attach state
- Removes commented out lines
- Refactors several WS methods
- Reuses code snippets
- Protects attached state
- Fixes a couple of negation bugs

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexis.com>

rpastrana commented:
It's prob not common for transactional ESP services to connect to dali, but the ESP process hosting those services could connect to dali.
@afishbeck can you provide more details about the request for this feature?

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: f1edb8cee86747117baecac5af3c44d4b5ed7311
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |84 |84 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |759 |759 |0 |
|test (thor) |686 |686 |0 |
|test (roxie) |799 |799 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   10 sec (00:00:10) |  184 sec (00:03:04) |   56 sec (00:00:56) |    6 sec (00:00:06) |   21 sec (00:00:21) | 1129 sec (00:18:49) |   17 sec (00:00:17) | 1423 sec (00:23:43) |

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: e359696c2d271c7bcddebf7fde41cf2c1157d4dc
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |84 |84 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |759 |759 |0 |
|test (thor) |686 |686 |0 |
|test (roxie) |799 |799 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   24 sec (00:00:24) |  183 sec (00:03:03) |   56 sec (00:00:56) |    6 sec (00:00:06) |   22 sec (00:00:22) | 1156 sec (00:19:16) |   17 sec (00:00:17) | 1464 sec (00:24:24) |

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 046f464bf8211e8db5f598f8a83480b957a85ce6
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |85 |85 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |759 |759 |0 |
|test (thor) |686 |686 |0 |
|test (roxie) |799 |799 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   12 sec (00:00:12) |  184 sec (00:03:04) |   56 sec (00:00:56) |    6 sec (00:00:06) |   21 sec (00:00:21) | 1118 sec (00:18:38) |   16 sec (00:00:16) | 1413 sec (00:23:33) |

richardkchapman commented:
Why is the wselk commit on this branch?

rpastrana commented:
It shouldn't be, I will remove to correct branch

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: e359696c2d271c7bcddebf7fde41cf2c1157d4dc
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |85 |85 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |759 |759 |0 |
|test (thor) |689 |689 |0 |
|test (roxie) |802 |802 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    5 sec (00:00:05) |  183 sec (00:03:03) |   56 sec (00:00:56) |    6 sec (00:00:06) |   21 sec (00:00:21) | 1111 sec (00:18:31) |   16 sec (00:00:16) | 1398 sec (00:23:18) |

afishbeck commented:
@rpastrana there are still a couple outstanding comments I made previously.. otherwise looks fine.

rpastrana committed:
HPCC-15037 Add mechanism to detach ESP from DALI

- Rewords error message

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexis.com>

rpastrana commented:
@afishbeck please see latest commit and replies. Thanks.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: d2bc14b3fb3e1774a0cd68ea509486f695be2f02
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |88 |88 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |776 |776 |0 |
|test (thor) |707 |707 |0 |
|test (roxie) |820 |820 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   34 sec (00:00:34) |  183 sec (00:03:03) |   56 sec (00:00:56) |    6 sec (00:00:06) |   22 sec (00:00:22) | 1181 sec (00:19:41) |   17 sec (00:00:17) | 1499 sec (00:24:59) |

afishbeck commented:
@richardkchapman @rpastrana in regards to the WHY? questions... static transactional ESPs don't need to connect to DALI, but DynamicESDL (the future transactional ESP for most cases) gets its dynamic content from DALI.  Detaching it from DALI acts in much the same way detaching roxie does.  The content is then fixed until reattached, and any other side effects of being connected to DALI (if any) are avoided.

We support a simple unsubscribe and a full detach.  The full detach only really exists because of historical fears of being attached to DALI in production.

afishbeck reviewed (approved):
@rpastrana looks fine.
